### ISSUE TITLE
Expose API like `pulumi.runtime.setMocks` to mock `Config` and `StackReferences`

### ISSUE THREAD

**Issue by zbuchheit** [pulumi/pulumi#16042]
## Hello!

- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

### Ask from user

Improve upon the unit testing experience in Pulumi by exposing an API similar to `pulumi.runtime.setMocks` to mock `Config` and `StackReferences`. 

### Additional Details

While it is possible to mock these resources today, it isn't easy. There has been some previous discussion about this experience [here](https://github.com/pulumi/pulumi/issues/9212) as well as overall testing experience feedback in other issues.

#### Highlighted pain points

- You can mock calls to `newResource` and `call` via `pulumi.runtime`, but not `StackReferences` and `Config` objects. These are required to be mocked or the test will not conclude. 
- Mocking library choices can cause many differences in implementation
- You need a stubbing library to accomplish the mocking.
  - Must return from stubbed functions
  - Pass the correct nesting level of stubbed functions
  - Verify that stubbed functions are being called

### Affected area/feature

Testing SDK

### Related issues 
https://github.com/pulumi/pulumi/issues/10975
https://github.com/pulumi/pulumi/issues/4472
https://github.com/pulumi/pulumi/issues/9212

**Comment by mikhailshilkov** [comment]
@zbuchheit I would prefer to keep separate issues for Config vs Stack References. In fact, these look like #4472 and #9212 respectively. Is this issue asking for anything beyond those?

Also, #9212 is closed as done. [Here](https://github.com/pulumi/pulumi/blob/master/sdk/nodejs/tests/stackReference.spec.ts) is a test example. There was a comment about .NET not working, but it looks like this has been resolved in pulumi-dotnet too. Can you summarize what is still missing?
